### PR TITLE
[ROCm] dynamically hipify CUDA EP, ROCm EP is now a shared provider

### PR DESCRIPTION
**Pull request by jeffdaily**
Due to the semi-static nature of the ROCm EP, it was not up to date with other CUDA EP improvements.  The ROCm EP is now hipified at build from the CUDA EP.  Some ROCm files that were not suitable for hipify-at-build were re-hipified and changes committed back to source control.

Migrating as much as possible to hipify-at-build will help prevent duplicate efforts between the ROCm and CUDA EPs.  Additional operators are suitable for hipification, dependent on these changes.

### CODE REVIEW

File: include/onnxruntime/core/framework/allocator.h [hunk 1]
```diff
@@ -37,6 +37,8 @@ namespace onnxruntime {
 constexpr const char* CPU = "Cpu";
 constexpr const char* CUDA = "Cuda";
 constexpr const char* CUDA_PINNED = "CudaPinned";
+constexpr const char* ROCM = "Rocm";
+constexpr const char* ROCM_PINNED = "RocmPinned";
```
Comment (weixingzhang) [inline include/onnxruntime/core/framework/allocator.h:41]: Let's not add new type for now. I had a PR to rename CUDA/CUDA_PINNED to GPU/GPU_PINNED and share it between CUDA and ROCm. For some weird failure, the PR couldn't be merged. Whenever I have cycles, I will re-visit that PR again.

Comment (jeffdaily) [reply include/onnxruntime/core/framework/allocator.h]: Fixed by commit 0ae9ef56c4a53f7f664b18780ee192b8d4acf0e5.

**Comment by jeffdaily** [comment]
@weixingzhang or @suffiank , do the azure pipelines automatically rerun after new commits are pushed?  I had to fix the build for CUDA as well as address @weixingzhang's commit to remove ROCM_PINNED.

**Comment by jeffdaily** [comment]
@suffiank should we have seen CI status results by now?

**Comment by weixingzhang** [comment]
> Due to the semi-static nature of the ROCm EP, it was not up to date with other CUDA EP improvements. The ROCm EP is now hipified at build from the CUDA EP. Some ROCm files that were not suitable for hipify-at-build were re-hipified and changes committed back to source control.
> 
> Migrating as much as possible to hipify-at-build will help prevent duplicate efforts between the ROCm and CUDA EPs. Additional operators are suitable for hipification, dependent on these changes.

Hi Jeff, thanks for the contribution! I saw you have put if (USE_ROCM) in CUDA files, which I was trying not to do to make CUDA files clean. Initially, I was in dilemma: duplicate some codes vs share more code with CUDA by using USE_ROCM. Conservatively, I chose the former way to make the files clean. For this part of changes, can you separate them in a different PR? This PR just focusing on making ROCm EP as shared provider.

**Comment by jeffdaily** [comment]
> > Due to the semi-static nature of the ROCm EP, it was not up to date with other CUDA EP improvements. The ROCm EP is now hipified at build from the CUDA EP. Some ROCm files that were not suitable for hipify-at-build were re-hipified and changes committed back to source control.
> > Migrating as much as possible to hipify-at-build will help prevent duplicate efforts between the ROCm and CUDA EPs. Additional operators are suitable for hipification, dependent on these changes.
> 
> Hi Jeff, thanks for the contribution! I saw you have put if (USE_ROCM) in CUDA files, which I was trying not to do to make CUDA files clean. Initially, I was in dilemma: duplicate some codes vs share more code with CUDA by using USE_ROCM. Conservatively, I chose the former way to make the files clean. For this part of changes, can you separate them in a different PR? This PR just focusing on making ROCm EP as shared provider.

The duplication of some codes is what contributed to the large delta between the CUDA and ROCM EPs.  I am in favor of the former approach (USE_ROCM in CUDA files).

I admit this is a large PR.  I can attempt to split it into shared provider vs dynamically hipifying additional operators.  

Alternatively, this PR (if passing CI) could be accepted as-is today, then we walk back the insertions of USE_ROCM into the CUDA files with follow-up PRs?  That might prove to be less work than attempting to separate this PR into multiple.

**Comment by jeffdaily** [comment]
@weixingzhang I have made a pass through the PR to identify where I've added or removed USE_ROCM.  I have determined that all 14 uses of USE_ROCM that skip native bfloat16 data types can be removed if I update the hipify script to leave `CUDA_VERSION` alone.  In any case, it is incorrect to hipify `CUDA_VERSION` -> `ROCM_VERSION` or `__CUDA_ARCH__` -> `__ROCM_ARCH__` because the semantic meaning of these values are different between CUDA and ROCM.

Here is a detailed file by file breakdown.  19 total USE_ROCM are unavoidable additions to cuda EP files.  IMHO this is a reasonable cost to avoid further maintenance headaches in the future.

- onnxruntime/core/providers/cpu
  - cpu_provider_shared.cc
    - [1] NECESSARY: next to existing USE_CUDA to declare Size__Compute et al
  - cpu_provider_shared.h
    - [1] NECESSARY: next to existing USE_CUDA to implement Size__Compute et al
    - [1] NECESSARY: next to existing USE_CUDA to implement GatherElements et al
- onnxruntime/core/providers/cuda
  - cu_inc/common.cuh
    - [2] bfloat16
    - [4] half intrinsics are unconditionally available on ROCM
    - [4] __shfl_* intrinsics instead of __shfl_*_sync on ROCM
  - cuda_call.cc
    - [1] rocblas CASE_TO_ENUM
  - cuda_common.h
    - [1] bfloat16
    - [1] CublasMathModeSetter not needed for ROCM
  - cuda_execution_provider.cc
    - [1] define NO_ROCM, then use NO_ROCM later
    - [179] NO_ROCM used
    - [4] bfloat16
    - [2] RNNNeedFallbackToCPU, ConvTransposeNeedFallbackToCPU conditionally defined and used, instead LSTM/RNN/GRU/ConvTranspose not supported on ROCM
  - cuda_execution_provider.h
    - [1] GetCudnnConvAlgo is undefined for ROCM
    - [2] bfloat16
  - cuda_utils.cu
    - [2] bfloat16
  - math/softmax.cc
    - [2] call miopenSoftmaxForward_V2 instead of cudnnSoftmaxForward
    - [2] don't implement double data type
    - [2] bfloat16
  - math/softmax_impl.cu
    - [1] WARP_BATCH = 1
    - [1] batches_per_warp = 1 and threads_per_block = 256
    - [1] bfloat16
- onnxruntime/core/providers/shared_library:
  - provider_bridge_provider.cc
    - [1] NECESSARY: next to existing USE_CUDA to enable TileOp::IsTileMemcpy et al
  - provider_interfaces.h
    - [1] NECESSARY: mirrors CUDA block above, adding ROCM allocator et al
    - [1] NECESSARY: next to existing USE_CUDA to enable PhiloxGenerator
- onnxruntime/core/session/provider_bridge_ort.cc
    - [1] NECESSARY: mirrors CUDA block above, adding NCCL support for ROCM
    - [1] NECESSARY: mirrors CUDA block above, implementing ROCM allocator et al
    - [1] NECESSARY: next to existing USE_CUDA to implement PhiloxGenerator
    - [1] NECESSARY: mirrors CUDA block above, adding NCCL support for ROCM

- places where USE_ROCM was **removed**
  - [-1] onnxruntime/core/providers/shared_library/provider_api.h
  - [-1] onnxruntime/core/session/onnxruntime_c_api.cc
  - [-1] onnxruntime/test/framework/test_utils.cc
  - [-2] onnxruntime/test/framework/test_utils.h
  - [-1] onnxruntime/test/util/default_providers.cc
  - [-1] orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

**Comment by jeffdaily** [comment]
A follow-up PR that depends on this PR will add ops TopK, NonMaxSuppression, GatherND, Resize, and Upsample.  There are additional ops efforts underway.

**Comment by jeffdaily** [comment]
@weixingzhang would you be able to re-review after my latest changes?

**Comment by jeffdaily** [comment]
Closing in favor of #8877 .

**Pull request closed without merging**